tool/thingtalk-dataset.js

```diff
@@ -0,0 +1,312 @@
+        this._bar = null;
+        this._i = 0;
+        
+        if (this.locale.toLowerCase() == 'zh-tw') {
+            this._opencc_t2s = new OpenCC('t2s.json');
+            this._opencc_s2t = new OpenCC('s2t.json');
+        }
+    }
+
+    async _transform(ex, encoding, callback) {
+        if (ex.type == 'meta') {
+            this._language = ex.data.language;
+            this._bar = new ProgressBar('[ :bar ] :current/:total', { total: ex.data.num_examples });
+        } else if (ex.type == 'example') {
+            try {
+                ex.data.preprocessed = await this.preprocessAll(ex.data.utterances, ex.data.id);
+            } catch (e) {
+                console.log(ex.data.id);
+                console.log(ex.data.utterances);
+                console.log(ex.data.preprocessed);
+                throw e;
+            }
+            this._i += 1;
+            this._bar.tick({ current: this._i });
+        }
+        callback(null, ex);
+    }
+    
+    async preprocessAll(utterances, id) {
+        const promises = await utterances.map(async utterance => {
+            if (this.locale.toLowerCase() == 'zh-tw')
+                utterance = this._opencc_t2s.convertSync(utterance)
+            let preprocessed = await tokenizeExample(this._tokenizer, utterance, id, this._language);
+            if (this.locale.toLowerCase() == 'zh-tw')
+                preprocessed = this._opencc_s2t.convertSync(preprocessed)
+            return preprocessed;
+        });
+        const preprocessedAll = await Promise.all(promises);
+        return preprocessedAll;
+    }
+
+    _flush(callback) {
+        this._tokenizer.end();
+        process.nextTick(callback);
+    }
+}
+
+class ThingTalkDatasetReader extends stream.Readable {
+    constructor(options) {
+        super({ objectMode: true });
+        this._tpClient = options.thingpediaClient;
+        
+        this._options = options;
+        
+        this._initialization = null;
+        this._dataset = null;
+        this._i = null;
+    }
+    
+    _read() {
+        if (this._initialization === null)
+            this._initialization = this._initialize();
+
+        this._initialization.then(() => this._output()).catch((e) => {
+            console.error(e);
+            this.emit('error', e);
+        });
+    }
+    
+    _output() {
+        if (this._i == this._dataset.examples.length) {  // done
+            this.push(null);
+            return;
+        } else if (this._i == -1) {
+            this.push({
+                type: 'meta',
+                data: {
+                    name: this._dataset.name,
+                    language: this._dataset.language,
+                    annotations: this._dataset.annotations,
+                    num_examples: this._dataset.examples.length
+                }
+            });
+        } else {
+            this.push({
+                type: 'example',
+                data: this._dataset.examples[this._i]
+            });
```

Comment:
Pushing two different things in the same stream smells of bad design. Maybe streams are not the right abstraction here? After all, your streams don't stream at all, everything is cached in memory at once.

Reply:
1. I put type and data here because I need to hold the meta data of the dataset as well to reconstruct dataset.tt. I search before but nodejs does not have streaming with meta data officially.
2. It's not streaming because dataset.prettyprint() is not streaming.

So...

Should I either 1) not to user streaming when processing tt dataset, or 2) write a streaming version of prettyprint()?

Reply:
A streaming version of prettyprint() could be useful (for example in almond-cloud to implement APIs that return datasets more efficiently), but also not quite easy.
What I mean though, is that, if you need metadata, and metadata is not supported between streams, maybe nodejs streams/pipe are not the right abstraction and you should not use streams at all. Anyway this is more of style/readability comment than a hard requirement, maybe we'll clean it up in the future.